ddtrace/settings.py

```diff
@@ -129,13 +189,57 @@ def __init__(self, global_config, *args, **kwargs):
         self.global_config = global_config
         self.hooks = Hooks()
         self.http = HttpConfig()
+        self._integration_name = integration_name
+
+    def __getattribute__(self, key):
+        item = super(IntegrationConfig, self).__getattribute__(key)
+        if key == '_attrs':
+            return item
+        return item if key in self._attrs else item.value
+
+    def __getitem__(self, key):
+        item = super(IntegrationConfig, self).__getitem__(key)
+        if isinstance(item, IntegrationConfigItem):
+            return item.value
+        else:
+            return item
+
+    def __setattr__(self, key, value):
+        if key == '_attrs':
+            return super(IntegrationConfig, self).__setattr__(key, value)
+
+        # Ignore self attributes that we don't want to encapsulate with an
+        # IntegrationConfigItem.
+        if key in self._attrs:
+            return super(IntegrationConfig, self).__setattr__(key, value)
+
+        if not hasattr(super(IntegrationConfig, self), key):
+            item = IntegrationConfigItem(self, key)
+        else:
+            # Have to use the super getattribute since ours will unwrap the
+            # IntegrationConfigValue.
+            item = super(IntegrationConfig, self).__getattribute__(key)
+        item.update_value(value)
+        item = super(IntegrationConfig, self).__setattr__(key, item)
+        return item.value if item else item
 
     def __deepcopy__(self, memodict=None):
-        new = IntegrationConfig(self.global_config, deepcopy(dict(self)))
+        new = IntegrationConfig(self.global_config, self.integration_name, deepcopy(dict(self)))
         new.hooks = deepcopy(self.hooks)
         new.http = deepcopy(self.http)
         return new
 
+    def _add_default(self, key, value):
+        """Adds a default key and value to the configuration.
+
+        Note: for internal integration usage only.
+        :param key: key of the item
+        :param value: value of the item
+        :return:
+        """
+        item = IntegrationConfigItem(self, key, default_value=value)
```

Comment:
We should add a check to see if this key exists yet, if it exists then we need to do an update on the config item, or just not set this as the default (or something)

Reply:
I added a warning below since calling `_add_default` on an existing key should not happen. I don't think we should do an update on the config item since that will flag it as being user-modified and will take precedence.